### ISSUE TITLE
Update `TimePicker` component design

### ISSUE THREAD

**Issue by jameskoster** [WordPress/gutenberg#64390]
Extracted from https://github.com/WordPress/gutenberg/pull/64267#discussion_r1710061434

[`TimePicker`](https://wordpress.github.io/gutenberg/?path=/docs/components-timepicker--docs) was originally designed with a fairly narrow scope; the date picking UI in the page inspector:

<img width="378" alt="Screenshot 2024-08-09 at 09 56 49" src="https://github.com/user-attachments/assets/5a1853c2-508e-441a-9674-76dbb2e10988">

In this context (popover) there is room enough for a more elaborate UI with multiple controls. However in a more traditional form setting this design is convoluted and overly prominent. Let's explore a more lightweight design, applicable either as a variant or a full redesign. 

Inspiration can be taken from the [`datetime-local` input type](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/datetime-local) which condenses everything into a single text field with a built-in calendar/time selection popover: 

<img width="363" alt="Screenshot 2024-08-09 at 09 59 11" src="https://github.com/user-attachments/assets/c7b65faf-7e44-4d93-a8e2-d99f2c8ae630">

**Comment by mirka** [comment]
Possibly related: #60971

**Comment by seifeldinio** [comment]
Here's a suggested design update for the `TimePicker` component.  
In the new design, I’ve taken into consideration the need for a more lightweight and streamlined UI.

Here's a prototype to showcase the updated design in action:

https://github.com/user-attachments/assets/71296395-606f-495c-a60a-db4b530179e6

Component:

![component](https://github.com/user-attachments/assets/335774f6-9aa3-4525-a8ca-f575ff995bb6)


Mockup:

![mockup](https://github.com/user-attachments/assets/8787cbfd-347a-4811-83db-9b301a2be3c0)


Would love to hear your thoughts!

**Comment by christopherabalos** [comment]
While I agree that TimePicker feels heavy, the separate controls also ensure accessibility. Calendar and time controls are two of the most difficult UIs to make accessible.

The native datetime-location input type allow selection of each date/time segment:

<img width="363" alt="356532746-c7b65faf-7e44-4d93-a8e2-d99f2c8ae630" src="https://github.com/user-attachments/assets/91572134-044b-40aa-a8e9-6e2e4046a1ab">

Something similar would need to happen in an updated TimePicker.

**Comment by jameskoster** [comment]
@christopherabalos absolutely, 100%! I think the first step here is to assess the accessibility of the `datetime-location` input type, and how much of that should be embodied 1:1 in the updated design. 

From there it'll mostly be a case of applying the wp design language to the visuals, as @seifeldinio did above. 

It's worth keeping date _range_ picking in mind as well (https://github.com/WordPress/gutenberg/issues/60971 as pointed out above), with additional shortcuts for things like "Last week", "Month to date", etc. This will be useful for filtering data views (e.g. to show pages published in the last month).

**Comment by mirka** [comment]
Might be stupid question, but what is stopping us from just using the native `<input type="datetime-local">` (or `<InputControl type="datetime-local">`)?  Especially when we want to shift to this "popover" style. A [recent accessibility test](https://www.digitala11y.com/input-type-date-the-accessibility-of-html-date-picker/) said that "its overall experience surpasses any other custom date picker I’ve encountered on various booking platforms". And we get a ton of stuff for free, like mobile support.

Even for the date range picker case, I find it likely that the easiest way to build that accessibly is to use two native `datetime-local` inputs, rather than going for a custom or library solution.

**Comment by DaniGuardiola** [comment]
@mirka I would push for using the platform for these things every day of the week, for those same reasons. Unless there's a very specific, very critical requirement or constraint to not do that, I feel like these days it's the best option in almost every aspect, at the cost of slightly less design freedom.

**Comment by jameskoster** [comment]
I lean in that direction as well, and had the same thought about combining two native `datetime-local` inputs for the range picker. A detail missing there would be shortcuts for quick selecting presets like 'Month to date', 'Last week', etc., but that can be discussed in https://github.com/WordPress/gutenberg/issues/60971 and a design solution seems trivial on the surface.

As Dani points out the main drawback is styling; making it blend seamlessly with the broader UI may not be possible. That is arguably an acceptable trade-off though; there are too many benefits to ignore. The mobile experience in particular is really slick... getting all of this for free, without having to worry about the ongoing maintenance feels like a huge win:

<img width="1886" alt="Frame 1000005746" src="https://github.com/user-attachments/assets/1f394098-8443-469e-9701-0f81a7ae6723">

**Comment by DaniGuardiola** [comment]
@jameskoster glad to see some agreement! FWIW, there are some ways to still get the design closer to what we want, mainly:

- https://developer.mozilla.org/en-US/docs/Web/CSS/accent-color
- https://developer.mozilla.org/en-US/docs/Web/CSS/color-scheme

Not sure if they apply to these specific input types though. I would assume they do or, at least, they will at some point.

**Comment by jameskoster** [comment]
Assuming we adopt the native input the design for this one seems relatively straight-forward, largely taking cues from other inputs types:

<img width="1019" alt="Screenshot 2024-09-23 at 13 35 42" src="https://github.com/user-attachments/assets/e940dbbf-f6b6-46e6-ac31-9a080e2820cf">

There is however a question about implementation. In the post/page publish flows upcoming events are (loosely) displayed by virtue of an instance of `DatePicker` and its `events` prop. An equivalent would not be possible using the native input, so we'd need a different way of displaying upcoming events. Perhaps something like the following (left is current, right would be new):

<img width="1252" alt="date" src="https://github.com/user-attachments/assets/f14e1629-5b52-45d0-8b0e-6b3a5d2406b7">

Theoretically there could be a middle-ground:

<img width="1252" alt="middle" src="https://github.com/user-attachments/assets/70f48a50-269f-4c64-b73f-f8a55576069e">

What do y'all think?

**Comment by oandregal** [comment]
Thinking about [this use case](https://github.com/WordPress/gutenberg/issues/55100#issuecomment-1877128587) (filters for DataViews), for dates we'd want a mix of timepicker component + preset dates (today, last week, etc.). Is that a pattern that we'd use in more places than the DataViews filters? How do we see those two things working together (separate components, single one, etc.)?

**Comment by jameskoster** [comment]
@oandregal in addition to the simple `TimePicker` (an existing component) I think we need to create a new component for date range picking (which will likely have applications outside of filtering data views). ~I'll open a new issue~ never mind [we already have one](https://github.com/WordPress/gutenberg/issues/60971) :)
